core/types/graph.h

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include "core/list.h"
+#include "core/reference.h"
+
+class GraphVertex;
+class GraphEdge;
+
+struct GraphData {
+	Map<GraphVertex *, List<GraphEdge *>> data;
+	void remove_vertex(GraphVertex *p_vertex);
+	void remove_edge(GraphEdge *p_vertex);
+};
+
+class Graph : public Reference {
+	GDCLASS(Graph, Reference);
+
+	GraphData *graph = nullptr;
+
+protected:
+	static void _bind_methods();
+
+	GraphEdge *_add_edge(const Variant &a, const Variant &b, real_t weight, bool directed);
+
+public:
+	// Vertex API
+	GraphVertex *add_vertex(const Variant &p_value);
+	void remove_vertex(GraphVertex *p_vertex);
+	bool has_vertex(GraphVertex *p_vertex) const { return graph->data.has(p_vertex); }
+	Array get_vertex_list() const;
+	int get_vertex_count() const { return graph->data.size(); }
+
+	Array get_neighbors(GraphVertex *p_vertex);
+	Array get_successors(GraphVertex *p_vertex);
+	Array get_predecessors(GraphVertex *p_vertex);
+
+	// Edge API
+	GraphEdge *add_edge(const Variant &a, const Variant &b, real_t weight = 1.0);
+	GraphEdge *add_directed_edge(const Variant &a, const Variant &b, real_t weight = 1.0);
```

Comment:
`GraphVertex` has ability to assign attributes via `Variant value`, I think the same should be done to `GraphEdge`. This way, you can assign any label/attribute to both vertices and edges.

```suggestion
	GraphEdge *add_edge(const Variant &a, const Variant &b, const Variant &value = 1.0);
	GraphEdge *add_directed_edge(const Variant &a, const Variant &b, const Variant &value = 1.0);
```